internal/controller/driver_controller.go

```diff
@@ -17,46 +17,1211 @@ limitations under the License.
+									},
+									Resources: ptr.Deref(
+										pluginSpec.Resources.Liveness,
+										corev1.ResourceRequirements{},
+									),
+								})
+							}
+						}
+						return containers
+					}),
+					Volumes: utils.Call(func() []corev1.Volume {
+						volumes := append(
+							// Add user defined volumes at the start to make sure they do not
+							// overwrite built in volumes.
+							utils.MapSlice(
+								pluginSpec.Volumes,
+								func(v csiv1a1.VolumeSpec) corev1.Volume {
+									return v.Volume
+								},
+							),
+							utils.HostDevVolume,
+							utils.HostSysVolume,
+							utils.HostRunMountVolume,
+							utils.LibModulesVolume,
+							utils.KeysTmpDirVolume,
+							utils.PluginDirVolume(pluginSpec.KubeletDirPath, r.driverPrefix),
+							utils.PluginMountDirVolume(pluginSpec.KubeletDirPath),
+							utils.PodsMountDirVolume(pluginSpec.KubeletDirPath),
+							utils.RegistrationDirVolume(pluginSpec.KubeletDirPath),
+						)
+						if ptr.Deref(pluginSpec.EnableSeLinuxHostMount, false) {
+							volumes = append(
+								volumes,
+								utils.EtcSelinuxVolume,
+							)
+						}
+						if r.driver.Spec.Encryption != nil {
+							volumes = append(
+								volumes,
+								utils.KmsConfigVolume(&r.driver.Spec.Encryption.ConfigMapRef),
+							)
+						}
+						if r.isRdbDriver() {
+							volumes = append(
+								volumes,
+								utils.OidcTokenVolume,
+							)
+						}
+						return volumes
+					}),
+				},
+			},
+		}
+
+		return nil
+	})
+
+	logCreateOrUpdateResult(log, "node plugin daemonset", daemonSet, opResult, err)
+	return err
+}
+
+func (r *driverReconcile) reconcileLivnessService() error {
+	if r.driver.Spec.Liveness == nil {
+		return nil
+	}
+
+	service := &corev1.Service{}
+	service.Namespace = r.driver.Namespace
+	service.Name = r.generateName("livness")
+
+	log := r.log.WithValues("service", service.Name)
+	log.Info("Reconciling livness service")
+
+	opResult, err := ctrlutil.CreateOrUpdate(r.ctx, r.Client, service, func() error {
+		if err := ctrlutil.SetControllerReference(&r.driver, service, r.Scheme); err != nil {
+			log.Error(err, "Failed setting an owner reference on service")
+			return err
+		}
+
+		service.Spec = corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "csi-http-metrics",
+					Port:       8080,
+					Protocol:   corev1.ProtocolTCP,
+					TargetPort: intstr.FromInt(r.driver.Spec.Liveness.MetricsPort),
+				},
+			},
+		}
+
+		return nil
+	})
+
+	logCreateOrUpdateResult(log, "livness service", service, opResult, err)
+	return err
+}
+
+func (r *driverReconcile) isRdbDriver() bool {
+	return r.driverType == RbdDriverType
+}
+
+func (r *driverReconcile) isCephFsDriver() bool {
+	return r.driverType == CephFsDriverType
+}
+
+func (r *driverReconcile) isNfsDriver() bool {
+	return r.driverType == NfsDriverType
+}
+
+func (r *driverReconcile) generateName(suffix string) string {
+	return fmt.Sprintf("%s-%s", r.driver.Name, suffix)
```

Comment:
we need to have length check on the driver CR name to keep it minimal, we can do this as followup as well.

Reply:
Yes, this will go into the type definition itself, which I will have a PR for soon. 
I was not planning to add code for that